driver/controller.go

```diff
@@ -0,0 +1,134 @@
+package driver
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/Juniper/contrail-go-api"
+	"github.com/Juniper/contrail-go-api/types"
+)
+
+const (
+	DomainName = "default-domain"
+)
+
+type Info struct {
+}
+
+type Controller struct {
+	ApiClient contrail.ApiClient
+}
+
+func NewController(ip string, port int) *Controller {
+	client := &Controller{}
+	client.ApiClient = contrail.NewClient(ip, port)
+	return client
+}
+
+func (c *Controller) GetNetwork(tenantName, networkName string) (*types.VirtualNetwork,
+	error) {
+	name := fmt.Sprintf("%s:%s:%s", DomainName, tenantName, networkName)
+	net, err := types.VirtualNetworkByName(c.ApiClient, name)
+	if err != nil {
+		return nil, err
+	}
+	return net, nil
+}
+
+func (c *Controller) GetDefaultGatewayIp(net *types.VirtualNetwork) (string, error) {
+	ipamReferences, err := net.GetNetworkIpamRefs()
+	if err != nil {
+		return "", err
+	}
+	if len(ipamReferences) == 0 {
+		return "", errors.New("Ipam references list is empty")
+	}
+	attribute := ipamReferences[0].Attr
+	ipamSubnets := attribute.(types.VnSubnetsType).IpamSubnets
+	if len(ipamSubnets) == 0 {
+		return "", errors.New("Ipam subnets list is empty")
+	}
+	gw := ipamSubnets[0].DefaultGateway
+	if gw == "" {
+		return "", errors.New("Default GW is empty")
+	}
+	return gw, nil
+}
+
+func (c *Controller) GetOrCreateInstance(tenantName, containerId string) (*types.VirtualMachine, error) {
+	name := fmt.Sprintf("%s:%s:%s", DomainName, tenantName, containerId)
+	instance, err := types.VirtualMachineByName(c.ApiClient, name)
+	if err == nil && instance != nil {
+		return instance, nil
+	}
+
+	instance = new(types.VirtualMachine)
+	instance.SetFQName("project", []string{DomainName, tenantName, containerId})
+	err = c.ApiClient.Create(instance)
+	if err != nil {
+		return nil, err
+	}
+	return instance, nil
+}
+
+func (c *Controller) GetOrCreateInterface(net *types.VirtualNetwork,
+	instance *types.VirtualMachine) (*types.VirtualMachineInterface, error) {
+	instanceFQName := instance.GetFQName()
+	namespace := instanceFQName[len(instanceFQName)-2]
+	name := fmt.Sprintf("%s:%s:%s", DomainName, namespace, instance.GetName())
+	iface, err := types.VirtualMachineInterfaceByName(c.ApiClient, name)
+	if err == nil && iface != nil {
+		return iface, nil
+	}
+
+	iface = new(types.VirtualMachineInterface)
+	iface.SetFQName("project", []string{DomainName, namespace, instance.GetName()})
+	err = iface.AddVirtualMachine(instance)
+	if err != nil {
+		return nil, err
+	}
+	err = iface.AddVirtualNetwork(net)
+	if err != nil {
+		return nil, err
+	}
+	err = c.ApiClient.Create(iface)
+	if err != nil {
+		return nil, err
+	}
+	return iface, nil
+}
+
+func (c *Controller) GetInterfaceMac(iface *types.VirtualMachineInterface) (string, error) {
+	macs := iface.GetVirtualMachineInterfaceMacAddresses()
+	if len(macs.MacAddress) == 0 {
+		return "", errors.New("Empty MAC list")
+	}
+	return macs.MacAddress[0], nil
```

Comment:
https://github.com/Juniper/contrail-controller/blob/master/src/schema/vnc_cfg.xsd#L1528
it can have more MACs

but I think 1 is "normal" scenario, and in case of containers, they have 1 MAC